### PR TITLE
How to deprecate a recipe option

### PR DESCRIPTION
**Pull request by uilianries**
Sometimes we need to change recipes options, one of most intuitive action is removing deprecated options, but it could cause a break change if you are using that option in your project. Instead, on CCI we follow an undocumented protocol. Let's add it on FAQ.

/cc  @SpaceIm @ericLemanissier 

- [ ] I've read the [guidelines](https://github.com/conan-io/conan-center-index/blob/master/docs/how_to_add_packages.md) for contributing.
- [ ] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [ ] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [ ] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by Croydon** [comment]
We only had brief discussions about this in some pull requests. Does the majority of core contributors think that this is the best way? 

Personally, I think that we break recipes in so many ways so often, that this way of deprecating options might have a good intention, but is ultimately almost pointless. The only way people can prevent the risk of breakages is by using revisions and if you use them, no option can suddenly disappear. If you then try to update revisions Conan will inform you that the option does not exist.

Removing old options from recipes would help to keep them cleaner.

Also, if consumers set options they expect that they actually have an effect. If recipes just output some warning they might be easily overlooked in several hundred or thousand lines of log output. Assuming that options have an effect, getting no error from Conan but then noticing maybe much later that the options had no effect is IMHO the much worse experience.

**Comment by prince-chrismc** [comment]
> we break recipes in so many ways so often [...] the only way people can prevent the risk of breakages is by using revisions

:100:

Any consumer who needs consistency is

- cloning all the recipes and packages 
- using revisions

> and if you use them, no option can suddenly disappear

My only mute counter is it makes it that much harder to upgrade. However if deprecating an option is a breaking change then it's no worse.

File: docs/faqs.md
Comment (prince-chrismc) [inline docs/faqs.md:126]: It feels a bit circular, with the answer in the question

```suggestion
## Can I remove an option from recipe
```

**Comment by SSE4** [comment]
I am not sure if I like it... 
1. it's so much extra code
2. we keep an option which has no effect
3. we only print a warning, and usually attention is not paid for them

also I don't get it. you say:
```
No. Changing any option will result in a different package ID and may result a different behavior, the result can break users.
```
but then you suggest:
```
def package_id(self):
    del self.info.options.foobar
```
it means if I had `-o foobar=True` in my profile, package id will change anyway for that configuration, right?

**Comment by uilianries** [comment]
> it means if I had -o foobar=True in my profile, package id will change anyway for that configuration, right?

Yes. Actually, the real problem is when you that option and we remove it from a recipe. When updating it, or consuming a fresh remote copy, Conan command will break because that option is no longer available.

**Comment by uilianries** [comment]
I see many comments on this PR, which is excellent, more people giving their opinion.

Deprecating options is a suggestion, not a rule, but as we are discussing now, we can provide cases where we should remove or not an option.

File: docs/faqs.md
Comment (SSE4) [inline docs/faqs.md:128]: please remove that argument, as adding `del self.info.options.foobar` causes the same effect as removing option (different package ID), so there is no advantage for that case

Comment (uilianries) [inline docs/faqs.md:128]:
```suggestion
No. Removing any option will result in a missing value, it can break any user who is using it.
```

**Review by prince-chrismc — APPROVED** [review]
I think it's good to merge our current policy but I'd love to see reform in the future

**Comment by ericLemanissier** [comment]
disabling the effect of an option instead of removing the option from the recipe, is choosing a silent breakage over a loud breakage. This does not seem to be the better choice to me either.

**Comment by uilianries** [comment]
Thank you all for your comments! After discussing with Conan Team about this topic, we see removing options is better, because we are faking when add "deprecated", and no all people read that warning message. I'll update this PR, following the opposite side.

**Comment by SpaceIm** [comment]
Don't you remember what happened with libcurl?

**Comment by uilianries** [comment]
Yes, I do.

Deprecating options can be a remedy for cases like libcurl, but also we are lying for users. We patched an option which is no longer. The recipe is turning a patchwork.

Conan Center Index can be improved, rebuilding what is needed to avoid missing package IDs after removing options.

Lockfiles are the key for behavior and reproducibility. So even if we update a package, a Lockfile will keep the same package.

**Comment by SpaceIm** [comment]
> Deprecating options can be a remedy for cases like libcurl, but also we are lying for users. We patched an option which is no longer.

We don't lie, the old options are working as expected.

**Comment by uilianries** [comment]
> We don't lie, the old options are working as expected.

Options are working for old revisions only. After deprecating we can pass same old values, but there is no effect for recipe.

> I agree, but how many consumers are using lockfiles? Theory vs reality?

Companies are strongly recommended to use Lockfiles. Regular users can update their package revisions, as soon as Conan Center regenerates all packages affected by an option change.

**Comment by SpaceIm** [comment]
> > We don't lie, the old options are working as expected.
> 
> Options are working for old revisions only. After deprecating we can pass same old values, but there is no effect for recipe.

I tried to carefully map the deprecated options values to the corresponding new option values:

```python
        # Deprecated options
        # ===============================
        if (any(deprecated_option != "deprecated" for deprecated_option in [self.options.with_openssl, self.options.with_wolfssl, self.options.with_winssl, self.options.darwin_ssl])):
            self.output.warn("with_openssl, with_winssl, darwin_ssl and with_wolfssl options are deprecated. Use with_ssl option instead.")
            if tools.is_apple_os(self.settings.os) and self.options.with_ssl == "darwinssl":
                if self.options.darwin_ssl == True:
                    self.options.with_ssl = "darwinssl"
                elif self.options.with_openssl == True:
                    self.options.with_ssl = "openssl"
                elif self.options.with_wolfssl == True:
                    self.options.with_ssl = "wolfssl"
                else:
                    self.options.with_ssl = False
            if not tools.is_apple_os(self.settings.os) and self.options.with_ssl == "openssl":
                if self.settings.os == "Windows" and self.options.with_winssl == True:
                    self.options.with_ssl = "schannel"
                elif self.options.with_openssl == True:
                    self.options.with_ssl = "openssl"
                elif self.options.with_wolfssl == True:
                    self.options.with_ssl = "wolfssl"
                else:
                    self.options.with_ssl = False
        # ===============================
```

**Comment by prince-chrismc** [comment]
I think `libcurl` is likely an exception, we changed from many settings to one... but the "setting" still exists. For that use case I think `"deprecated"` was the right call.

I think it much more common we have "broken" or settings that do not work, removed from upstream, etc.
For example I recently did this...

https://github.com/conan-io/conan-center-index/blob/1c747532270cd01bdad45f3be693f8d9f7fc55c7/recipes/tomlplusplus/all/conanfile.py#L37-L38

but the setting _never_ worked, no one was using it nor was it's logic maintained.... it could have been removed.

**Comment by jgsogo** [comment]
I think we should prefer the _loud breakage_ over the _silent_ one: 

> disabling the effect of an option instead of removing the option from the recipe, is choosing a silent breakage (I compiled my package successfully, but it does not have the feature I asked for), over a loud breakage (conan signals from the start that there is a problem). This does not seem to be the better choice to me either.

It can be very misleading if you are setting an option in your profile or command line and it has no effect... it can lead to some hours of wtf debugging and investigation 😅 

It is also true that a deprecation like [the one did by @SpaceIm](https://github.com/conan-io/conan-center-index/pull/5020#issuecomment-806809503) would be the better approach, but sometimes it adds lot's of complexity to a recipe that we will want to remove at some point in time. Maybe the policy should follow these lines:
 * Prefer the _depreation_ path with a mapping from old options to new ones.
 * If logic is too complex (this is subjective and depends on the PR author) then just remove the option.
 * After some time (comment with a date?), we will welcome a PR removing the option that was deprecated.

I'm not a friend of deferring the inevitable, but I'll be ok with the two steps deprecation as well.

**Comment by prince-chrismc** [comment]
The only thing I would suggest changing is

```diff
-this is subjective and depends on the PR author
+this is subjective and depends on the Conan team
```

**Review by a4z — APPROVED** [review]

**Review by SSE4 — DISMISSED** [review]

**Comment by uilianries** [comment]
Thank you all for your comments! I made a small update, I've included Javier's lines, now deprecation is not mandatory, but a strong recommendation. An option can be removed after 1 month of deprecation. Of course, we need to check possible gaps (downstream breakage) before merging anything.

File: docs/faqs.md
Comment (prince-chrismc) [inline docs/faqs.md:128]:
```suggestion
It's preferable to leave all options (ie. not remove them) because it may break other packages which require those deleted options.
```

**Review by prince-chrismc — COMMENTED** [review]
Minor editorial on the double negation

**Review by prince-chrismc — APPROVED** [review]

**Review by a4z — APPROVED** [review]

**Review by SSE4 — APPROVED** [review]

**Review by danimtb — APPROVED** [review]

**Pull request merged**
